tests/MongoDB.Driver.Tests/ClientSessionHandleTests.cs

```diff
@@ -298,5 +773,40 @@ public static IServerSession _serverSession(this ClientSessionHandle obj)
             var fieldInfo = typeof(ClientSessionHandle).GetField("_serverSession", BindingFlags.NonPublic | BindingFlags.Instance);
             return (IServerSession)fieldInfo.GetValue(obj);
         }
+
+        // TransactionExecutor
+        public static int __transactionTimeout()
+        {
+            var fieldInfo = typeof(TransactionExecutor).GetField("__transactionTimeout", BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Static);
+            return (int)((TimeSpan)fieldInfo.GetValue(null)).TotalSeconds;
+        }
+
+        public static bool CommitWithRetries(this ClientSessionHandle obj, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var methodInfo = typeof(TransactionExecutor).GetMethod("CommitWithRetries", BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Static);
+                return (bool)methodInfo.Invoke(obj, new object[] { obj, startTime, clock, cancellationToken });
+            }
+            catch (Exception ex)
+            {
+                // Reflection wraps a method exception into TargetInvocationException
+                throw (ex as TargetInvocationException)?.InnerException ?? ex;
+            }
+        }
+
+        public static Task<bool> CommitWithRetriesAsync(this ClientSessionHandle obj, DateTime startTime, IClock clock, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var methodInfo = typeof(TransactionExecutor).GetMethod("CommitWithRetriesAsync", BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Static);
+                return (Task<bool>)methodInfo.Invoke(obj, new object[] { obj, startTime, clock, cancellationToken });
+            }
+            catch (Exception ex)
+            {
+                // Reflection wraps a method exception into TargetInvocationException
+                throw (ex as TargetInvocationException)?.InnerException ?? ex;
+            }
+        }
```

Comment:
Note: above refactoring will require changes to Reflector.cs. The following existing methods need to be refactored to throw the inner exception:

        public static object InvokeStatic(Type type, string name, BindingFlags flags = BindingFlags.NonPublic | BindingFlags.Static)
        {
            var methodInfo = type.GetMethods(flags)
                .Where(m => m.Name == name && m.GetParameters().Length == 0)
                .Single();
            try
            {
                return methodInfo.Invoke(null, new object[] { });
            }
            catch (TargetInvocationException exception)
            {
                throw exception.InnerException;
            }
        }

        public static object InvokeStatic<T1>(Type type, string name, T1 arg1)
        {
            var parameterTypes = new[] { typeof(T1) };
            var methodInfo = type.GetMethods(BindingFlags.NonPublic | BindingFlags.Static)
                .Where(m => m.Name == name && m.GetParameters().Select(p => p.ParameterType).SequenceEqual(parameterTypes))
                .Single();
            try
            {
                return methodInfo.Invoke(null, new object[] { arg1 });
            }
            catch (TargetInvocationException exception)
            {
                throw exception.InnerException;
            }
        }

Reply:
The following new method will be needed in Reflector.cs:

    public static object InvokeStatic<T1, T2, T3, T4>(Type type, string name, T1 arg1, T2 arg2, T3 arg3, T4 arg4)
    {
        var parameterTypes = new[] { typeof(T1), typeof(T2), typeof(T3), typeof(T4) };
        var methodInfo = type.GetMethods(BindingFlags.NonPublic | BindingFlags.Static)
            .Where(m => m.Name == name && m.GetParameters().Select(p => p.ParameterType).SequenceEqual(parameterT
            .Single();
        try
        {
            return methodInfo.Invoke(null, new object[] { arg1, arg2, arg3, arg4 });
        }
        catch (TargetInvocationException exception)
        {
            throw exception.InnerException;
        }
    }

Reply:
For consistency go ahead and add the 2 and 3 arg versions of InvokeStatic at the same time.

Reply:
>BUT... why is a class called ClientSessionHandleTests testing methods in the TransactionExecutor?

The idea was to have `extension` methods instead of direct calling of static classes. 

Done